src/models/Satisfiability.jl

```diff
@@ -0,0 +1,181 @@
+"""
+Basic useful functions for Boolean algebra (https://github.com/QuEraComputing/GenericTensorNetworks.jl/blob/master/src/networks/Satisfiability.jl)
+"""
+
+"""
+    BoolVar{T}
+    BoolVar(name, neg)
+
+Boolean variable for constructing CNF clauses.
+"""
+struct BoolVar{T}
+    name::T
+    neg::Bool
+end
+BoolVar(name) = BoolVar(name, false)
+function Base.show(io::IO, b::BoolVar)
+    b.neg && print(io, "¬")
+    print(io, b.name)
+end
+
+"""
+    CNFClause{T}
+    CNFClause(vars)
+
+A clause in [`CNF`](@ref), its value is the logical or of `vars`, where `vars` is a vector of [`BoolVar`](@ref).
+"""
+struct CNFClause{T}
+    vars::Vector{BoolVar{T}}
+end
+function Base.show(io::IO, b::CNFClause)
+    print(io, join(string.(b.vars), " ∨ "))
+end
+Base.:(==)(x::CNFClause, y::CNFClause) = x.vars == y.vars
+Base.length(x::CNFClause) = length(x.vars)
+
+"""
+    CNF{T}
+    CNF(clauses)
+
+Boolean expression in [conjunctive normal form](https://en.wikipedia.org/wiki/Conjunctive_normal_form).
+`clauses` is a vector of [`CNFClause`](@ref), if and only if all clauses are satisfied, this CNF is satisfied.
+
+Example
+------------------------
+Under development
+"""
+struct CNF{T}
+    clauses::Vector{CNFClause{T}}
+end
+function Base.show(io::IO, c::CNF)
+    print(io, join(["($k)" for k in c.clauses], " ∧ "))
+end
+Base.:(==)(x::CNF, y::CNF) = x.clauses == y.clauses
+Base.length(x::CNF) = length(x.clauses)
+function is_kSAT(cnf::CNF)
+    len_lst = [length(clause.vars) for clause in cnf.clauses]
+    if length(unique(len_lst)) == 1
+        return len_lst[1]
+    else
+        return false
```

Comment:
Return value not type stable: https://docs.julialang.org/en/v1/manual/performance-tips/